From 73ac32a35d45790fa2b6b66808c16fbc549930fb Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Fri, 31 Oct 2025 15:18:45 +0800
Subject: [PATCH 3/5] mtk: mt76: mt7996: offload radar threshold initialization
 to fw

Offload radar threshold initialization to fw
Add rdd_radar_th, rdd_pulse_th, & rdd_fcc5_lpn debugfs file for user to
set customized radar specs

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/debugfs.c | 96 ++++++++++++++++++++++++++++++++++++++++++++++++
 mt7996/mac.c     | 77 --------------------------------------
 mt7996/mac.h     |  5 ---
 3 files changed, 96 insertions(+), 82 deletions(-)

diff --git a/mt7996/debugfs.c b/mt7996/debugfs.c
index 43d9c5771..54dd17303 100644
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -1234,6 +1234,99 @@ mt7996_airtime_read(struct seq_file *s, void *data)
 	return 0;
 }
 
+static ssize_t mt7996_rdd_radar_th_set(struct file *file,
+				       const char __user *user_buf,
+				       size_t count, loff_t *ppos)
+{
+	struct mt7996_dev *dev = file->private_data;
+	struct mt7996_dfs_pattern p;
+	char buf[100];
+	int ret, radar_type;
+
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	if (copy_from_user(buf, user_buf, count))
+		return -EFAULT;
+
+	if (count && buf[count - 1] == '\n')
+		buf[count - 1] = '\0';
+	else
+		buf[count] = '\0';
+
+	if (sscanf(buf,
+		   "%d %hhu %hhu %hhu %hhu %hhu %hhu %u %u %hhu %hhu %hhu %hhu %hhu %hhu %u",
+		   &radar_type, &p.enb, &p.stgr, &p.min_crpn, &p.max_crpn, &p.min_crpr,
+		   &p.min_pw, &p.min_pri, &p.max_pri, &p.max_pw, &p.min_crbn, &p.max_crbn,
+		   &p.min_stgpn, &p.max_stgpn, &p.min_stgpr, &p.min_stgpr_diff) != 16) {
+		dev_err(dev->mt76.dev, "Invalid format\n");
+		return -EINVAL;
+	}
+
+	ret = mt7996_mcu_set_radar_th(dev, radar_type, &p);
+	if (ret)
+		return ret;
+	return count;
+}
+
+static const struct file_operations fops_rdd_radar_th = {
+	.write = mt7996_rdd_radar_th_set,
+	.open = simple_open,
+	.owner = THIS_MODULE,
+	.llseek = default_llseek,
+};
+
+static ssize_t mt7996_rdd_pulse_th_set(struct file *file,
+				       const char __user *user_buf,
+				       size_t count, loff_t *ppos)
+{
+	struct mt7996_dev *dev = file->private_data;
+	struct mt7996_dfs_pulse p;
+	char buf[100];
+	int ret;
+
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	if (copy_from_user(buf, user_buf, count))
+		return -EFAULT;
+
+	if (count && buf[count - 1] == '\n')
+		buf[count - 1] = '\0';
+	else
+		buf[count] = '\0';
+
+	if (sscanf(buf,
+		   "%u %d %d %u %u %u %u",
+		   &p.max_width, &p.max_pwr, &p.min_pwr, &p.min_stgr_pri,
+		   &p.max_stgr_pri, &p.min_cr_pri, &p.max_cr_pri) != 7) {
+		dev_err(dev->mt76.dev, "Invalid format\n");
+		return -EINVAL;
+	}
+
+	ret = mt7996_mcu_set_pulse_th(dev, &p);
+	if (ret)
+		return ret;
+	return count;
+}
+
+static const struct file_operations fops_rdd_pulse_th = {
+	.write = mt7996_rdd_pulse_th_set,
+	.open = simple_open,
+	.owner = THIS_MODULE,
+	.llseek = default_llseek,
+};
+
+static int mt7996_rdd_fcc5_lpn_set(void *data, u64 val)
+{
+	struct mt7996_dev *dev = data;
+
+	return mt7996_mcu_set_fcc5_lpn(dev, val);
+}
+
+DEFINE_DEBUGFS_ATTRIBUTE(fops_rdd_fcc5_lpn, NULL,
+			 mt7996_rdd_fcc5_lpn_set, "%lld\n");
+
 int mt7996_init_band_debugfs(struct mt7996_phy *phy)
 {
 	struct mt7996_dev *dev = phy->dev;
@@ -1264,6 +1357,9 @@ int mt7996_init_band_debugfs(struct mt7996_phy *phy)
 				    &fops_radar_trigger);
 		debugfs_create_devm_seqfile(dev->mt76.dev, "rdd_monitor", dir,
 					    mt7996_rdd_monitor);
+		debugfs_create_file("rdd_radar_th", 0200, dir, dev, &fops_rdd_radar_th);
+		debugfs_create_file("rdd_pulse_th", 0200, dir, dev, &fops_rdd_pulse_th);
+		debugfs_create_file("rdd_fcc5_lpn", 0200, dir, dev, &fops_rdd_fcc5_lpn);
 	}
 
 #ifdef CONFIG_MTK_DEBUG
diff --git a/mt7996/mac.c b/mt7996/mac.c
index 392d4d838..a6f965027 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -14,45 +14,6 @@
 #include "vendor.h"
 #include "mt7996_trace.h"
 
-static const struct mt7996_dfs_radar_spec etsi_radar_specs = {
-	.pulse_th = { 110, -10, -80, 40, 5200, 128, 5200 },
-	.radar_pattern = {
-		[5] =  { 1, 0,  6, 32, 28, 0,  990, 5010, 17, 1, 1 },
-		[6] =  { 1, 0,  9, 32, 28, 0,  615, 5010, 27, 1, 1 },
-		[7] =  { 1, 0, 15, 32, 28, 0,  240,  445, 27, 1, 1 },
-		[8] =  { 1, 0, 12, 32, 28, 0,  240,  510, 42, 1, 1 },
-		[9] =  { 1, 1,  0,  0,  0, 0, 2490, 3343, 14, 0, 0, 12, 32, 28, { }, 126 },
-		[10] = { 1, 1,  0,  0,  0, 0, 2490, 3343, 14, 0, 0, 15, 32, 24, { }, 126 },
-		[11] = { 1, 1,  0,  0,  0, 0,  823, 2510, 14, 0, 0, 18, 32, 28, { },  54 },
-		[12] = { 1, 1,  0,  0,  0, 0,  823, 2510, 14, 0, 0, 27, 32, 24, { },  54 },
-	},
-};
-
-static const struct mt7996_dfs_radar_spec fcc_radar_specs = {
-	.pulse_th = { 110, -10, -80, 40, 5200, 128, 5200 },
-	.radar_pattern = {
-		[0] = { 1, 0,  8,  32, 28, 0, 508, 3076, 13, 1,  1 },
-		[1] = { 1, 0, 12,  32, 28, 0, 140,  240, 17, 1,  1 },
-		[2] = { 1, 0,  8,  32, 28, 0, 190,  510, 22, 1,  1 },
-		[3] = { 1, 0,  6,  32, 28, 0, 190,  510, 32, 1,  1 },
-		[4] = { 1, 0,  9, 255, 28, 0, 323,  343, 13, 1, 32 },
-	},
-};
-
-static const struct mt7996_dfs_radar_spec jp_radar_specs = {
-	.pulse_th = { 110, -10, -80, 40, 5200, 128, 5200 },
-	.radar_pattern = {
-		[0] =  { 1, 0,  8,  32, 28, 0,  508, 3076,  13, 1,  1 },
-		[1] =  { 1, 0, 12,  32, 28, 0,  140,  240,  17, 1,  1 },
-		[2] =  { 1, 0,  8,  32, 28, 0,  190,  510,  22, 1,  1 },
-		[3] =  { 1, 0,  6,  32, 28, 0,  190,  510,  32, 1,  1 },
-		[4] =  { 1, 0,  9, 255, 28, 0,  323,  343,  13, 1, 32 },
-		[13] = { 1, 0,  7,  32, 28, 0, 3836, 3856,  14, 1,  1 },
-		[14] = { 1, 0,  6,  32, 28, 0,  615, 5010, 110, 1,  1 },
-		[15] = { 1, 1,  0,   0,  0, 0,   15, 5010, 110, 0,  0, 12, 32, 28 },
-	},
-};
-
 static void mt7996_scan_rx(struct mt7996_phy *phy);
 static void mt7996_rx_beacon_hint(struct mt7996_phy *phy, struct mt7996_vif *mvif);
 
@@ -3420,40 +3381,6 @@ static int mt7996_dfs_start_radar_detector(struct mt7996_phy *phy)
 	return err;
 }
 
-static int
-mt7996_dfs_init_radar_specs(struct mt7996_phy *phy)
-{
-	const struct mt7996_dfs_radar_spec *radar_specs;
-	struct mt7996_dev *dev = phy->dev;
-	int err, i;
-
-	switch (dev->mt76.region) {
-	case NL80211_DFS_FCC:
-		radar_specs = &fcc_radar_specs;
-		err = mt7996_mcu_set_fcc5_lpn(dev, 8);
-		if (err < 0)
-			return err;
-		break;
-	case NL80211_DFS_ETSI:
-		radar_specs = &etsi_radar_specs;
-		break;
-	case NL80211_DFS_JP:
-		radar_specs = &jp_radar_specs;
-		break;
-	default:
-		return -EINVAL;
-	}
-
-	for (i = 0; i < ARRAY_SIZE(radar_specs->radar_pattern); i++) {
-		err = mt7996_mcu_set_radar_th(dev, i,
-					      &radar_specs->radar_pattern[i]);
-		if (err < 0)
-			return err;
-	}
-
-	return mt7996_mcu_set_pulse_th(dev, &radar_specs->pulse_th);
-}
-
 int mt7996_dfs_init_radar_detector(struct mt7996_phy *phy)
 {
 	struct mt7996_dev *dev = phy->dev;
@@ -3473,10 +3400,6 @@ int mt7996_dfs_init_radar_detector(struct mt7996_phy *phy)
 		goto stop;
 
 	if (prev_state <= MT_DFS_STATE_DISABLED) {
-		err = mt7996_dfs_init_radar_specs(phy);
-		if (err < 0)
-			return err;
-
 		err = mt7996_dfs_start_radar_detector(phy);
 		if (err < 0)
 			return err;
diff --git a/mt7996/mac.h b/mt7996/mac.h
index a1205f3a4..cc9a6811b 100644
--- a/mt7996/mac.h
+++ b/mt7996/mac.h
@@ -37,11 +37,6 @@ struct mt7996_dfs_pattern {
 	u32 min_stgpr_diff;
 } __packed;
 
-struct mt7996_dfs_radar_spec {
-	struct mt7996_dfs_pulse pulse_th;
-	struct mt7996_dfs_pattern radar_pattern[16];
-};
-
 #define MT7996_SDO_EVENT_COUNT			GENMASK(26, 20)
 #define MT7996_SDO_EVENT_DW_LEN			GENMASK(31, 27)
 #define MT7996_SDO_EVENT_ID			GENMASK(26, 21)
-- 
2.45.2

